src/man/mod.rs

```diff
@@ -0,0 +1,373 @@
+mod author;
+mod environment;
+mod flag;
+mod option;
+
+use self::author::Author;
+use self::environment::Env;
+use self::flag::Flag;
+use self::option::Opt;
+use roff::{bold, italic, list, Roff, Troffable};
+use std::convert::AsRef;
+
+/// Man page struct.
+#[derive(Debug)]
+pub struct Man {
+  name: String,
+  description: Option<String>,
+  authors: Vec<Author>,
+  flags: Vec<Flag>,
+  options: Vec<Opt>,
+  environment: Vec<Env>,
+  arguments: Vec<String>,
+}
+
+impl Man {
+  /// Create a new instance.
+  pub fn new(name: &str) -> Self {
+    Self {
+      name: name.into(),
+      description: None,
+      authors: vec![],
+      flags: vec![],
+      options: vec![],
+      arguments: vec![],
+      environment: vec![],
+    }
+  }
+
+  /// Add a description.
+  pub fn description(mut self, desc: &str) -> Self {
+    let desc = desc.into();
+    self.description = Some(desc);
+    self
+  }
+
+  /// Add an author.
+  pub fn author(
+    mut self,
+    name: impl AsRef<str>,
+    email: Option<String>,
+  ) -> Self {
+    self.authors.push(Author {
+      name: name.as_ref().to_owned(),
+      email,
+    });
+    self
+  }
+
+  /// Add an environment variable.
+  pub fn environment(
+    mut self,
+    name: String,
+    default: Option<String>,
+    description: Option<String>,
+  ) -> Self {
+    self.environment.push(Env {
+      name,
+      default,
+      description,
+    });
+    self
+  }
+
+  /// Add an flag.
+  pub fn flag(
+    mut self,
+    short: Option<String>,
+    long: Option<String>,
+    description: Option<String>,
+  ) -> Self {
+    self.flags.push(Flag {
+      short,
+      long,
+      description,
+    });
+    self
+  }
+
+  /// Add an option.
+  pub fn option(
+    mut self,
+    short: Option<String>,
+    long: Option<String>,
+    description: Option<String>,
+    argument: String,
+    default: Option<String>,
+  ) -> Self {
```

Comment:
That's quite a lot of params for one function.

Looking at this it might make sense to make the underlying Option type public (but not its fields) and expose a builder patter for it, too. (You can probably use a derive).

Same for consistency's sake for the other similar methods.

Reply:
I agree, thinking something along the lines of https://github.com/rust-clique/man/issues/2#issuecomment-407015044 would be better.

I kind of want to merge the API so we can start working on demos; and then quickly follow it up with a separate API to get the naming and things closer to what's outlined in the issue above. Definitely agree this should be better!